**Project Akhir**

Nama : Mathius Octaroyhans Simanjuntak

NIM : 5311422076

**Memprediksi Pengguna yang berpindah pada Perusahaan Telekomunikasi**

Churn adalah pindahnya pelanggan dari satu provider ke provider lainnya. Perusahaan lebih memilih untuk mempertahankan customer, karena dibutuhkan biaya yang lebih sedikit daripada menambah customer yang baru.

Penjelasan setiap kolom <br>

•  Gender: Jenis kelamin pelanggan.<br>
•  SeniorCitizen: Status apakah pelanggan adalah warga lanjut usia atau bukan.<br>
•  Partner: Apakah pelanggan memiliki pasangan atau tidak.<br>
•  Dependents: Apakah pelanggan memiliki tanggungan atau tidak.<br>
•  MultipleLines: Apakah pelanggan memiliki lebih dari satu saluran atau tidak.<br>
•  InternetService: Jenis layanan internet yang digunakan oleh pelanggan.<br>
•  OnlineSecurity: Apakah pelanggan memiliki keamanan online atau tidak.<br>
•  OnlineBackup: Apakah pelanggan memiliki cadangan online atau tidak.<br>
•  DeviceProtection: Apakah pelanggan memiliki perlindungan perangkat atau tidak.<br>
•  TechSupport: Apakah pelanggan memiliki dukungan teknis atau tidak.<br>
•  StreamingTV: Apakah pelanggan memiliki layanan streaming TV atau tidak.<br>
•  StreamingMovies: Apakah pelanggan memiliki layanan streaming film atau tidak.<br>
•  Unnamed: 13: Sepertinya ini adalah kolom yang tidak memiliki label dan tidak memiliki nilai.<br>
•  Tenure: Lamanya pelanggan telah menggunakan layanan.<br>
•  PhoneService: Apakah pelanggan memiliki layanan telepon atau tidak.<br>
•  Contract: Jenis kontrak yang dimiliki pelanggan.<br>
•  PaperlessBilling: Apakah pelanggan menggunakan tagihan tanpa kertas atau tidak.<br>
•  PaymentMethod: Metode pembayaran yang digunakan oleh pelanggan.<br>
•  MonthlyCharges: Biaya bulanan yang dibebankan kepada pelanggan.<br>
•  TotalCharges: Total biaya yang dibebankan kepada pelanggan.<br>
•  Churn: Apakah pelanggan telah meninggalkan layanan atau tidak (churn).




## **IMPORT LIBRARY DAN DATA**

###**1. Import Library**

In [1]:
import numpy as np
import pandas as pd

Menggunakan 2 library, yaitu numpy dan pandas.NumPy adalah pustaka Python yang kuat untuk komputasi numerik. Ini menyediakan objek array multidimensi yang efisien, bersama dengan berbagai fungsi untuk bekerja dengan array ini. Pandas adalah pustaka Python yang populer untuk analisis dan manipulasi data. Ini dibangun di atas NumPy dan menyediakan struktur data tingkat tinggi, terutama DataFrame dan Series, yang memungkinkan pengguna untuk dengan mudah menyimpan, membersihkan, dan menganalisis data tabular.

###**2. Menampilkan informasi data**

Kita import dulu file csv nya yang bernama churn_data.csv

In [2]:
df = pd.read_csv("churn_data.csv")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7042 entries, 0 to 7041
Data columns (total 22 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7042 non-null   object 
 1   gender            7042 non-null   object 
 2   SeniorCitizen     7042 non-null   int64  
 3   Partner           7042 non-null   object 
 4   Dependents        7042 non-null   object 
 5   MultipleLines     7042 non-null   object 
 6   InternetService   7042 non-null   object 
 7   OnlineSecurity    7042 non-null   object 
 8   OnlineBackup      7042 non-null   object 
 9   DeviceProtection  7042 non-null   object 
 10  TechSupport       7042 non-null   object 
 11  StreamingTV       7042 non-null   object 
 12  StreamingMovies   7042 non-null   object 
 13  Unnamed: 13       0 non-null      float64
 14  tenure            7042 non-null   int64  
 15  PhoneService      7042 non-null   object 
 16  Contract          7042 non-null   object 


In [3]:
df.drop('Unnamed: 13', axis=1, inplace=True)

saya drop Unnamed :13 karena tidak ada isinya

###**3. Menampilkan 20 data teratas**

In [4]:
df.head(20)

customerID  gender  SeniorCitizen Partner Dependents     MultipleLines  \
0   7590-VHVEG  Female              0     Yes         No  No phone service   
1   5575-GNVDE    Male              0      No         No                No   
2   3668-QPYBK    Male              0      No         No                No   
3   7795-CFOCW    Male              0      No         No  No phone service   
4   9237-HQITU  Female              0      No         No                No   
5   9305-CDSKC  Female              0      No         No               Yes   
6   1452-KIOVK    Male              0      No        Yes               Yes   
7   6713-OKOMC  Female              0      No         No  No phone service   
8   7892-POOKP  Female              0     Yes         No               Yes   
9   6388-TABGU    Male              0      No        Yes                No   
10  9763-GRSKD    Male              0     Yes        Yes                No   
11  7469-LKBCI    Male              0      No         No                No   
12  8091-TTVAX    Male              0     Yes         No               Yes   
13  0280-XJGEX    Male              0      No         No               Yes   
14  5129-JLPIS    Male              0      No         No                No   
15  3655-SNQYZ  Female              0     Yes        Yes               Yes   
16  8191-XWSZG  Female              0      No         No                No   
17  9959-WOFKT    Male              0      No        Yes               Yes   
18  4190-MFLUW  Female              0     Yes        Yes                No   
19  4183-MYFRB  Female              0      No         No                No   

   InternetService       OnlineSecurity         OnlineBackup  \
0              DSL                   No                  Yes   
1              DSL                  Yes                   No   
2              DSL                  Yes                  Yes   
3              DSL                  Yes                   No   
4      Fiber optic                   No                   No   
5      Fiber optic                   No                   No   
6      Fiber optic                   No                  Yes   
7              DSL                  Yes                   No   
8      Fiber optic                   No                   No   
9              DSL                  Yes                  Yes   
10             DSL                  Yes                   No   
11              No  No internet service  No internet service   
12     Fiber optic                   No                   No   
13     Fiber optic                   No                  Yes   
14     Fiber optic                  Yes                   No   
15     Fiber optic                  Yes                  Yes   
16              No  No internet service  No internet service   
17     Fiber optic                  Yes                   No   
18             DSL                   No                   No   
19     Fiber optic                   No                  Yes   

       DeviceProtection  ...          StreamingTV      StreamingMovies tenure  \
0                    No  ...                   No                   No      1   
1                   Yes  ...                   No                   No     34   
2                    No  ...                   No                   No      2   
3                   Yes  ...                   No                   No     45   
4                    No  ...                   No                   No      2   
5                   Yes  ...                  Yes                  Yes      8   
6                    No  ...                  Yes                   No     22   
7                    No  ...                   No                   No     10   
8                   Yes  ...                  Yes                  Yes     28   
9                    No  ...                   No                   No     62   
10                   No  ...                   No                   No     13   
11  No internet service  ...  No internet servic

#**Mengubah Data**

Kita ubah kolom yang kita gunakan yaitu Partner, PhoneService, PaperlessBilling, Dependets, Churn menjadi type boolean

In [5]:
df[['Partner', 'Dependents', 'PhoneService', 'PaperlessBilling', 'Churn']] = df[['Partner', 'Dependents', 'PhoneService', 'PaperlessBilling', 'Churn']].replace({'Yes': True, 'No': False})

Kemudian kita cek apakah sudah berubah

In [6]:
df.dtypes

customerID           object
gender               object
SeniorCitizen         int64
Partner                bool
Dependents             bool
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
tenure                int64
PhoneService           bool
Contract             object
PaperlessBilling       bool
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                  bool
dtype: object

In [7]:
df.head(10)

customerID  gender  SeniorCitizen  Partner  Dependents     MultipleLines  \
0  7590-VHVEG  Female              0     True       False  No phone service   
1  5575-GNVDE    Male              0    False       False                No   
2  3668-QPYBK    Male              0    False       False                No   
3  7795-CFOCW    Male              0    False       False  No phone service   
4  9237-HQITU  Female              0    False       False                No   
5  9305-CDSKC  Female              0    False       False               Yes   
6  1452-KIOVK    Male              0    False        True               Yes   
7  6713-OKOMC  Female              0    False       False  No phone service   
8  7892-POOKP  Female              0     True       False               Yes   
9  6388-TABGU    Male              0    False        True                No   

  InternetService OnlineSecurity OnlineBackup DeviceProtection  ...  \
0             DSL             No          Yes               No  ...   
1             DSL            Yes           No              Yes  ...   
2             DSL            Yes          Yes               No  ...   
3             DSL            Yes           No              Yes  ...   
4     Fiber optic             No           No               No  ...   
5     Fiber optic             No           No              Yes  ...   
6     Fiber optic             No          Yes               No  ...   
7             DSL            Yes           No               No  ...   
8     Fiber optic             No           No              Yes  ...   
9             DSL            Yes          Yes               No  ...   

  StreamingTV StreamingMovies tenure  PhoneService        Contract  \
0          No              No      1         False  Month-to-month   
1          No              No     34          True        One year   
2          No              No      2          True  Month-to-month   
3          No              No     45         False        One year   
4          No              No      2          True  Month-to-month   
5         Yes             Yes      8          True  Month-to-month   
6         Yes              No     22          True  Month-to-month   
7          No              No     10         False  Month-to-month   
8         Yes             Yes     28          True  Month-to-month   
9          No              No     62          True        One year   

  PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0             True           Electronic check          29.85         29.85   
1            False               Mailed check          56.95        1889.5   
2             True               Mailed check          53.85        108.15   
3            False  Bank transfer (automatic)          42.30       1840.75   
4             True           Electronic check          70.70        151.65   
5             True           Electronic check          99.65         820.5   
6             True    Credit card (automatic)          89.10        1949.4   
7            False               Mailed check          29.75         301.9   
8             True           Electronic check         104.80       3046.05   
9            False  Bank transfer (automatic)          56.15       3487.95   

   Churn  
0  False  
1  False  
2   True  
3  False  
4   True  
5   True  
6  False  
7  False  
8   True  
9  False  

[10 rows x 21 columns]

## **Pembersihan data**

1. Membersihkan data yang terduplikat

Kita mengecek data dan menghilangkan data yang memiliki kesamaan di semua kolom

In [8]:
# menghitung jumlah data yang kosong
null_count = df.isnull().sum().sum()

# menghitung jumlah data yang sama
duplicate_count = df.duplicated().sum()

# melihat hasilnya
print("Number of null values:", null_count,
      "\nNumber of duplicate values:", duplicate_count,
      '\nOriginal data shape:', df.shape[0])

# menghilangkan data yang sama
df = df.drop_duplicates()
df.shape

Number of null values: 0 
Number of duplicate values: 0 
Original data shape: 7042


(7042, 21)

disini tidak ada yang duplikat


### 2. Menentukan kolom-kolom data yang akan digunakan

Sama seperti data yang kita ubah ke type boolean

In [9]:
# data yang akan digunakan
data_sub = ['Partner', 'Dependents', 'PhoneService', 'PaperlessBilling', 'Churn']
data = df[data_sub]

# menghitung jumlah churn
data['Churn'].value_counts()

Churn
False    5173
True     1869
Name: count, dtype: int64

Disini dimana nilai pelanggana yang berpindah sebanyak 1869 sesuai dengan nilai truenya.

# **logistic regression**

Disini kita akan membuat fungsi sigmoid, menghitung cost function, menghitung gradien, melakukan gradient descent, Menambahkan kolom 1 ke X untuk intercept, inisialisasi parameter theta, melakukan gradient descent, Membuat prediksi menggunakan parameter theta yang diperoleh, Menghitung akurasi model.

In [10]:
# Fungsi sigmoid
def sigmoid(z):
    return 1 / (1 + np.exp(-z))

# Fungsi untuk menghitung cost function
def cost_function(X, y, theta):
    m = len(y)
    h = sigmoid(np.dot(X, theta))
    cost = (-1 / m) * np.sum(y * np.log(h) + (1 - y) * np.log(1 - h))
    return cost

# Fungsi untuk menghitung gradien
def gradient(X, y, theta):
    m = len(y)
    h = sigmoid(np.dot(X, theta))
    grad = (1 / m) * np.dot(X.T, (h - y))
    return grad

# Fungsi untuk melakukan gradient descent
def gradient_descent(X, y, theta, alpha, num_iters):
    m = len(y)
    costs = []
    for i in range(num_iters):
        theta = theta - alpha * gradient(X, y, theta)
        cost = cost_function(X, y, theta)
        costs.append(cost)
    return theta, costs

# Menambahkan kolom 1 ke X untuk intercept
X = data[data_sub].drop('Churn', axis=1)
y = data['Churn']
X = np.c_[np.ones((len(X), 1)), X]

# Inisialisasi parameter theta
theta = np.zeros(X.shape[1])

# Melakukan gradient descent
alpha = 0.01  # Learning rate
num_iters = 1000  # Jumlah iterasi
theta, costs = gradient_descent(X, y, theta, alpha, num_iters)

# Membuat prediksi menggunakan parameter theta yang diperoleh
y_pred = sigmoid(np.dot(X, theta)) >= 0.5

# Menghitung akurasi model
accuracy = np.mean(y_pred == y) * 100
print(f"Akurasi model: {accuracy:.2f}%")

Akurasi model: 73.46%


Disini didapatkan hasil akurasi sebesar 73.46%. Dapat kita simpulkan bahwa akurasinya sudah termasuk bagus namun belum terlalu bagus.